### ISSUE TITLE
The token is not valid yet

### ISSUE THREAD

**Issue by rsajja79** [dmitry-viskov/pylti1.3#146]
pylti1p3.exception.LtiException: Can't decode id_token: The token is not yet valid (iat)

I'm getting this error. Everytime I try to launch an app from canvas, this is the error I'm getting. It was working for a day or so, I changed some code in launch that deals with nrps to get the role of the user, I'm getting this error.

**Comment by hmoffatt** [comment]
It means the computer clocks are not synchronized. Use ntp.

**Comment by rsajja79** [comment]
Use NTP on the server where this flask application is running?

**Comment by hmoffatt** [comment]
You need the clocks properly synchronised everywhere - your Canvas server and your flask application server. The error means that there is a difference in the clocks and hence the application can't trust the token.

**Comment by sdiemer** [comment]
Hello,

I'm facing a similar issue but with servers having correct NTP configurations.
The problem is that the LMS servers are using different NTP servers than the application servers.
The LMS NTP server is in advance of ~0.15 seconds and then when requests are fast enough, the application throws the error `The token is not yet valid (iat)`.
In my case, it is not possible to change the NTP servers used in all implied systems due to internal politics.
Could it be possible to pass the `verify_iat` setting to pyjwt ([source](https://github.com/jpadilla/pyjwt/issues/814)) ?

Regards

**Comment by hmoffatt** [comment]
You can set the verify options already, by calling `set_jwt_verify_options()` on your message_launch object.
https://github.com/dmitry-viskov/pylti1.3/blob/master/pylti1p3/message_launch.py#L254

Rather than disabling the iat check altogether, I think what is needed is to add some leeway when calling jwt.decode, at https://github.com/dmitry-viskov/pylti1.3/blob/master/pylti1p3/message_launch.py#L715

Solutions are 1. patch the library, 2. extended MessageLaunch to override `validate_jwt_signature` to pass the leeway parameter, or 3. disable the built-in `iat` validation and do your own.

**Comment by rsajja79** [comment]
I agree with the response provided above. I was able to resolve the issue by adding the leeway parameter to the MessageLaunch. This adjustment accommodates the minor time discrepancy between the LMS and application servers without disabling the iat check altogether.

**Comment by hmoffatt** [comment]
I think it would be helpful for MessageLaunch to have a method to set the leeway passed to jwt.decode (`set_jwt_verify_leeway` perhaps), like it has `set_jwt_verify_options`.

**Comment by sdiemer** [comment]
Thank you for your responses.
I searched for explanation on the "leeway" setting of pyjwt, it seems it is only for the expiration date:
https://pyjwt.readthedocs.io/en/latest/usage.html#expiration-time-claim-exp
I will have to make some tests to able to choose a solution among the list you provided.
Regards

**Comment by hmoffatt** [comment]
> I searched for explanation on the "leeway" setting of pyjwt, it seems it is only for the expiration date:

No, it is also applied to the `iat` validation, see the code: https://github.com/jpadilla/pyjwt/blob/master/jwt/api_jwt.py#L320

**Comment by sdiemer** [comment]
I finally got the time to work on this.
I added the method `set_jwt_leeway` in the `MessageLaunch` class to handle the definition of the parameter since it is not provided in JWT through the `options` dict.
You can cherry pick [my commit](https://github.com/UbiCastTeam/pylti1.3/commit/102b6a9d4d07153161da9b0b950596e825b20a8a#diff-2d8ec79354fe4991d1ea646b3a9c29ddf59a36b000cc31a34c9a8f09f46ec6dcR260) if you want to integrate this change in the project.

Thank you again for your help.

**Comment by rsajja79** [comment]
I started getting this error "pylti1p3.exception.LtiServiceException: HTTP response [https://uiowa.instructure.com/login/oauth2/token]: 400 - {"error":"invalid_request","error_description":"the 'iat' must not be in the future"}"


It was working till today. Then I started getting this must not be in the future error. I have a leeway of 120.

**Comment by hmoffatt** [comment]
> It was working till today. Then I started getting this must not be in the future error. I have a leeway of 120.

Do you mean you applied @sdiemer 's patch in order to set the leeway?

Is your system clock synchronized correctly? 120 seconds is a very large discrepancy.

**Comment by rsajja79** [comment]
I've had the leeway set even before like I commented on November 7th. 120 is just for testing. I set it to 30 now and still get the error. "

![Image](https://github.com/user-attachments/assets/b9eabb8c-356d-4e58-8d17-2ca06b123c1c)

**Comment by hmoffatt** [comment]
A clock synchronization issue would explain why the problem just started.

**Comment by rsajja79** [comment]
How do i disable the iat check? I've set the flag to false in set_jwt_verify_options. It doesnt seem to be doing anything, as I'm still getting the error.

**Comment by hmoffatt** [comment]
You should be able to call `set_jwt_verify_options({"verify_iat": False}` on your MessageLaunch instance.

**Comment by rsajja79** [comment]
![Image](https://github.com/user-attachments/assets/aa1be252-01a8-481c-83fa-5bf790a6d27f)

Would chaning the validate_jwt_signature in message_launch.py dont apply the flag?

**Comment by hmoffatt** [comment]
Wait, the error seems to be that Canvas is rejecting the `iat` from you, not the other way around. This is in a service call like AGS, NPRS etc. The problem is still the system clock.

**Comment by rsajja79** [comment]
Okay, thank you. I will try to reset the system clock in the morning.

**Comment by rsajja79** [comment]
You are right. There is something going on with the "message_launch.has_nrps()" or "message_launch.get_nrps()". I removed this form my code, and seems to fix the error. I will take a look at these functions and how its effecting the iat token. I will share my finidng here, if any.

**Comment by hmoffatt** [comment]
Calls to those services *generate* a token which is set to the tool consumer (Canvas). The problem is with the `iat` in the generated token, not the validation of the token received from Canvas.
